### PR TITLE
Windows 10 SDK install tests

### PR DESCRIPTION
**Pull request by mokagio**
Builds on top of #144. Initially, this was a dedicated PR because I expected a lot of Git noise by trial and error in CI. But as I kept refining the implementation for the Windows 10 SDK version file validation, I also introduced what I hope are other improvements to the Windows scripts.

- Adds validation of the `.windows-10-sdk-version` content before attempting to use said content as the version to install, including comparing against an allow-list of valid versions
- Replaces all `Write-Host` with `Write-Output`, because `Write-Host` is meant for messages such as user prompts
- Adds a `-DryRun` option to the Windows 10 SDK install script, mostly so it can be tested without doing a full install
- Adds a `SkipWindows10SDKInstallation` to the host setup script, just in case

Together with #144, I think this should be enough for now to ship a new plugin (major) version and begin DRYing the Electron apps.

## Next steps

- Take a look at the CI setup time and consider moving the tests into a single script if the overhead is considerable
- Keep working on #133 to make sure the script properly handle errors

---

- [x] I have considered if this change warrants release notes and have added them to the appropriate section in the `CHANGELOG.md` if necessary. — N.A.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Please add an entry in the <code>CHANGELOG.md</code> file to describe the changes made by this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: bin/install_windows_10_sdk.ps1 [hunk 1]
```diff
@@ -1,18 +1,41 @@
...
 
 $windowsSDKVersionFile = ".windows-10-sdk-version"
 if (-not (Test-Path $windowsSDKVersionFile)) {
   Write-Output "[!] No Windows 10 SDK version file found at $windowsSDKVersionFile."
   exit 1
 }
 
-$windows10SDKVersion = Get-Content $windowsSDKVersionFile
+$windows10SDKVersion = Get-Content $windowsSDKVersionFile `
+  | Where-Object { $_ -match '\S' }  # Remove empty lines
```
Comment (AliSoftware) [inline bin/install_windows_10_sdk.ps1]: Worth testing if this other (and arguably more idiomatic) implementation would work too:
```suggestion
$windows10SDKVersion = (Get-Content -TotalCount 1 $windowsSDKVersionFile).Trim()
```

Though tbh not even sure the `.Trim()` is strictly needed

File: bin/install_windows_10_sdk.ps1 [hunk 2]
```diff
@@ -1,18 +1,41 @@
...
+
+if ($windows10SDKVersion.Count -ne 1) {
+  Write-Output "[!] Invalid version file format."
+  Write-Output "Expected exactly one non-empty line, got:"
+  Write-Output ($windows10SDKVersion -join "`n")  # The join poperly formats multiple lines
```
Comment (AliSoftware) [inline bin/install_windows_10_sdk.ps1]:
```suggestion
  Write-Output ($windows10SDKVersion -join "`n")  # The join properly formats multiple lines
```

File: bin/install_windows_10_sdk.ps1 [hunk 3]
```diff
@@ -1,18 +1,41 @@
...
+  exit 1
+}
+
+if ($windows10SDKVersion -notmatch '^\d+$') {
+  Write-Output "[!] Invalid version file format."
+  Write-Output "Expected an integer, got: '$windows10SDKVersion'"
```
Comment (AliSoftware) [inline bin/install_windows_10_sdk.ps1]: I'm not well versed in how Windows10 SDK versioning works, so completely open question: are we sure it's guaranteed to be an integer? As opposed to versions like `13b4` or `12.7` or `123a` also being a possibility?

File: bin/install_windows_10_sdk.ps1 (hunk 2 above)
Comment (mokagio) [reply bin/install_windows_10_sdk.ps1]: Before accepting this suggestion, I'll test [the one above](https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/153#discussion_r1961399525), because if it works I'd say we can remove this check altogether.

I mean, it might be good to let the user know the added rubbish in `.windows-10-sdk-version` but  it might also be an annoying build failure given we were able to parse it anyway.

File: bin/install_windows_10_sdk.ps1 (hunk 3 above)
Comment (mokagio) [reply bin/install_windows_10_sdk.ps1]: As far as I understand, yes. That's because the integer used in the file is then passed to the SDK downloader and used to identify the version according to the spec in the screenshot below, sourced from [here](https://learn.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2022):

image

Having said that, when I tried 18362, nothing happened:

image

I wonder whether that's because the version is 10.0.18362.1, there's something wrong with that version, or my command.

For the record, I tried passing 18362.1 and it still did not result in the SDK being installed.

This all makes me wonder whether we might be better off using a default SDK version in the installer, and only falling back to the version file and/or a parameter if the user provides one. After all, I implemented the version file because other tools use it and because the app I was working on had a check for the SDK being installed and so I wanted a single source of truth. But I'd argue now that our tooling takes care of the installing, that check is redundant: the build script should trust the SDK is available.

File: bin/install_windows_10_sdk.ps1 (hunk 1 above)
Comment (mokagio) [reply bin/install_windows_10_sdk.ps1]: > Though tbh not even sure the `.Trim()` is strictly needed

Turned out to be, but to account for leading and trailing whitespaces, not newlines. See 48410dbc98907c58e8c86cc1280c999ae5b35a24 and 85f7bc24116592482d15a9f6ddd8fc0b43975267

File: bin/install_windows_10_sdk.ps1 (hunk 3 above)
Comment (AliSoftware) [reply bin/install_windows_10_sdk.ps1]: I like the version file to enforce a specific version instead of always installing the latest though, especially for build reproducibility and idempotence (ie similar to a lockfile ensuring every rerun of a build on a same commits days later will still use the same libs and tools)

Comment (mokagio) [reply bin/install_windows_10_sdk.ps1]: Ah, good point 👍 

Given the timezone schedule that basically leaves me with a whole workday before the next round of feedback, I'll work on adding an allow-list for the valid versions. Otherwise, I would have suggested to merge this as is and ship the allow-list as a followup improvement.

File: bin/install_windows_10_sdk.ps1 [hunk 4]
```diff
@@ -1,17 +1,64 @@
+# Install the Windows 10 SDK and Visual Studio Build Tools using the value in .windows-10-sdk-version.
+#
+# The expected .windows-10-sdk-version format is a integer representing a valid SDK component id.
+# The list of valid component ids can be found at
+# https://learn.microsoft.com/en-us/visualstudio/install/workload-component-id-vs-build-tools?view=vs-2022
+#
+# Example:
+#
+#   20348
+
```
Comment (mokagio) [inline bin/install_windows_10_sdk.ps1:10]: See https://github.com/Automattic/a8c-ci-toolkit-buildkite-plugin/pull/144#discussion_r1961889583 by @iangmaia 

I added the documentation here rather than in the prepare script as per @iangmaia 's comment because it seemed like a better place where to explain the expectation.

Also, I am aware we have two PRs that add docs here, but it felt best to add a manual version in this context to get it over the line while waiting for the LLM-generated doc approach to be discussed, see #159 #160

**Comment by mokagio** [comment]
@iangmaia @AliSoftware I'd appreciate a second (third, fourth?) look after the latest followup, so ideally we can merge in #144 and keep going.

Thanks! 🙇‍♂️

File: .buildkite/pipeline.yml [hunk 5]
```diff
@@ -118,3 +118,98 @@ steps:
         notify:
           - github_commit_status:
               context: "pr_changed_files Tests: Edge Cases"
+
+  - group: ":windows: install_windows_10_sdk Tests"
```
Comment (AliSoftware) [inline .buildkite/pipeline.yml:122]: What about making use of our new `pr_changed_files` to run this group conditionally?

The nice thing is that this `a8c-ci-toolkit` repo is the only one for now where it'd actually be easy to use that helper in `shared-pipeline-vars` to set up an env var you could then use as an `if: build.env('HAS_INSTALL_WINDOWS_10_SDK_CHANGES') == 1`.

This is because unlike all other repos for which the `a8c-ci-toolkit` is not loaded on the `uploader` agent when the main step that does `pipeline upload` is run on it, our default steps don't include it as a plugin… in this specific repo we don't even need to load it as a plugin because it's already a binary in the working copy of the repo it's trying to run CI on 😛 

Which means you should be able to use `bin/pr_change_files` in this repo directly (instead of having the plugin loaded and having its `bin/` folder added to the `$PATH`) to `export HAS_INSTALL_WINDOWS_10_SDK_CHANGES=$(bin/pr_changed_files --stdout --any-match './bin/install_windows_10_sdk.ps1' './tests/test-install-windows-sdk.ps1')` 😉 

Same for the other one btw.

File: .buildkite/pipeline.yml [hunk 6]
```diff
@@ -118,3 +118,98 @@ steps:
...
+    steps:
+      - label: ":windows: install_windows_10_sdk Tests - Version file with valid format and version"
+        command: |
+          "20348" | Out-File .windows-10-sdk-version
+          .\tests\test-install-windows-10-sdk.ps1 -ExpectedExitCode 0
```
Comment (AliSoftware) [inline .buildkite/pipeline.yml:127]: I just realized the inconsistency we have here between `snake_case` used for `bin/*` scripts and `kebab-case` used for `tests/*` 🤔

File: bin/prepare_windows_host_for_app_distribution.ps1 [hunk 7]
```diff
@@ -8,24 +8,28 @@
 #  - Install the Windows 10 SDK if it detected a `.windows-10-sdk-version` file(2)
 #
 # (1) The certificate it installs is stored in our AWS SecretsManager storage (`windows-code-signing-certificate` secret ID)
-# (2) You can skip the Win10 install even if `.windows-10-sdk-version` file is present by using the `SKIP_WINDOWS_10_SDK_INSTALL=1` env var before calling this script
+# (2) You can skip the Windows 10 SDK installation regardless of whether `.windows-10-sdk-version` is present by calling the script with `-SkipWindows10SDKInstallation`.
```
Comment (AliSoftware) [inline bin/prepare_windows_host_for_app_distribution.ps1:11]: 👍

**Review by AliSoftware — APPROVED** [review]

File: .buildkite/pipeline.yml (hunk 5 above)
Comment (mokagio) [reply .buildkite/pipeline.yml:122]: That would be neat and great dogfooding.

File: .buildkite/pipeline.yml (hunk 6 above)
Comment (mokagio) [reply .buildkite/pipeline.yml:127]: I noticed that too, but didn't take the time to write it down or address it. Sounds like an easy follow up PR for me to tackle.
